### ISSUE TITLE
Reconsider document body icons and the current gutter view

### ISSUE THREAD

**Issue by balthisar** [shysaur/Fragaria#15]
Are you willing to sacrifice the in-body error buttons in favor of this:
- display errors _only_ in the gutter.
- eliminate the reference to "line number gutter" (now it's just the gutter)
- maintain the ability to toggle the gutter display on and off
- the gutter now has a property indicating whether or not it will display line numbers.
- can add a property to Fragaria (probably to the new error controller) that will turn the gutter on automatically if there are any current errors (user default preference).

I understand the point about the delegates in #13, but there's one crucial thing I didn't mention at the end of the issue report: it wouldn't work with the body icon.

Right now the errors buttons are implemented two distinct ways:
- In the highlighter. All errors in the visible range are drawn in the document body at once, and I can't think of a way that a delegate would be efficient here.
- Separately in the ruler. In the second case, error icons are added in the gutter only when the line is drawn, and really in the same manner as the breakpoints.

The gutter has no reason to know about errors; it only has to know what to draw. In this respect icons in the gutter are just another decoration, like the breakpoints, and this would be an appropriate use of a delegate if I understood your explanation correctly.

Previously in the document body if line wrap is turned off _and_ the lines are long, the warnings were not visible unless the user scrolls horizontally. I inserted a change to push them to the left of the text view to prevent this when line wrap is off, but the appearance on screen is ugly there ("better than nothing" is the only defense I can offer).

So my thought was to eliminate them from the text editor completely and display them _only_ in the gutter. Of course this would mean that the icons are not visible with line numbers off, but then again, neither are the breakpoint markers currently.

Looking at Xcode and most other text editors, the norm is that there's _always_ a gutter, and the option to display or not display line numbers happens within the gutter.

I won't make begin making a change this drastic unless you decide it's worth sacrificing the body error icon.

**Comment by shysaur** [comment]
For now, I won't remove in-body error buttons.

Maybe I'll reopen this issue after #10 is fixed.

Also, now that you mention it, I'll revert the "push errors to the left" change in the b496ac7 mega-commit because I see it as a bug, since it hides the code you need to fix more frequently than the old behavior, and working around it is a lot more difficult. Plus, the error highlighting is visible anyway (just the icon is hidden).

About delegates, I still disagree.

It's true that, in your implementation of gutter errors, buttons are added like breakpoints. But breakpoint adding is initiated by the view (because of an user event), not by the application, so my 3rd point (delegates are for view->controller->view feedback) still stands. How error icons are laid out/drawn is an implementation detail.

Consider the NSWindow example again: it's got subviews (in the gutter view, buttons are subviews), but when they are drawn, there is no method in NSWindowDelegate to tell the window how to draw the subviews, this is accomplished by subclassing the subviews. Plus, error buttons are not decorations (they can be clicked).

**Comment by shysaur** [comment]
> Looking at Xcode and most other text editors, the norm is that there's always a gutter, and the option to display or not display line numbers happens within the gutter.

This is a good observation. I'm opening a new issue for that.
